File: nara-frontend/src/api/diagnostic.ts
/**
 * Endpoints de diagnóstico.
 */
import { apiClient } from "./client";
import type {
  DiagnosticStartResponse,
  AnswerSubmitResponse,
  NextQuestionsResponse,
  EligibilityResponse,
  DiagnosticResultResponse,
  DiagnosticOwnerEmailResponse,
  DiagnosticCurrentStateResponse,
  MicroDiagnosticAnswerInput,
  MicroDiagnosticStartResponse,
  MicroDiagnosticStateResponse,
  MicroReportResponse,
} from "../types";

export interface StartDiagnosticPayload {
  email: string;
  full_name?: string;
  session_id?: string;
  consent_privacy: boolean;
  consent_marketing?: boolean;
  device_info?: Record<string, unknown>;
  utm_source?: string;
}

export interface SubmitAnswerPayload {
  question_id: number;
  question_text: string;
  question_area: string;
  answer_text?: string;
  answer_scale?: number;
  response_time_seconds?: number;
}

export async function startDiagnostic(
  payload: StartDiagnosticPayload
): Promise<DiagnosticStartResponse> {
  const { data } = await apiClient.post<DiagnosticStartResponse>(
    "/diagnostic/start",
    payload
  );
  return data;
}

export async function submitAnswer(
  diagnosticId: string,
  payload: SubmitAnswerPayload
): Promise<AnswerSubmitResponse> {
  const { data } = await apiClient.post<AnswerSubmitResponse>(
    `/diagnostic/${diagnosticId}/answer`,
    payload
  );
  return data;
}

export async function sendResumeLink(
  diagnosticId: string
): Promise<{ status: string; message: string }> {
  const { data } = await apiClient.post(
    `/diagnostic/${diagnosticId}/send-resume-link`
  );
  return data;
}

/** Timeout para geração de próximas perguntas (RAG + LLM). Backend pode levar 35–45s; margem extra evita erro no front. */
const NEXT_QUESTIONS_TIMEOUT_MS = 90_000;

export async function getNextQuestions(
  diagnosticId: string
): Promise<NextQuestionsResponse> {
  const { data } = await apiClient.get<NextQuestionsResponse>(
    `/diagnostic/${diagnosticId}/next-questions`,
    { timeout: NEXT_QUESTIONS_TIMEOUT_MS }
  );
  return data;
}

export async function checkEligibility(
  diagnosticId: string
): Promise<EligibilityResponse> {
  const { data } = await apiClient.get<EligibilityResponse>(
    `/diagnostic/${diagnosticId}/eligibility`
  );
  return data;
}

export async function finishDiagnostic(
  diagnosticId: string
): Promise<DiagnosticResultResponse> {
  const { data } = await apiClient.post<DiagnosticResultResponse>(
    `/diagnostic/${diagnosticId}/finish`
  );
  return data;
}

export async function getResultByToken(
  token: string
): Promise<DiagnosticResultResponse> {
  const { data } = await apiClient.get<DiagnosticResultResponse>(
    `/diagnostic/result/${token}`
  );
  return data;
}

export async function getOwnerEmailByToken(
  token: string
): Promise<DiagnosticOwnerEmailResponse> {
  const { data } = await apiClient.get<DiagnosticOwnerEmailResponse>(
    `/diagnostic/result/${token}/owner-email`
  );
  return data;
}

export async function getResultPdfByToken(token: string): Promise<Blob> {
  const res = await apiClient.get(`/diagnostic/result/${token}/pdf`, {
    responseType: "blob",
  });
  return res.data as Blob;
}

export async function getCurrentState(
  diagnosticId: string
): Promise<DiagnosticCurrentStateResponse> {
  const { data } = await apiClient.get<DiagnosticCurrentStateResponse>(
    `/diagnostic/${diagnosticId}/current-state`
  );
  return data;
}

export async function startMicroDiagnostic(
  token: string,
  area: string
): Promise<MicroDiagnosticStartResponse> {
  const { data } = await apiClient.post<MicroDiagnosticStartResponse>(
    `/diagnostic/result/${token}/micro-diagnostic/start`,
    { area }
  );
  return data;
}

export async function getMicroDiagnosticState(
  token: string,
  microId: string
): Promise<MicroDiagnosticStateResponse> {
  const { data } = await apiClient.get<MicroDiagnosticStateResponse>(
    `/diagnostic/result/${token}/micro-diagnostic/${microId}`
  );
  return data;
}

export async function submitMicroDiagnosticAnswers(
  token: string,
  microId: string,
  answers: MicroDiagnosticAnswerInput[]
): Promise<MicroDiagnosticStateResponse> {
  const { data } = await apiClient.post<MicroDiagnosticStateResponse>(
    `/diagnostic/result/${token}/micro-diagnostic/${microId}/submit`,
    { answers }
  );
  return data;
}

export async function finishMicroDiagnostic(
  token: string,
  microId: string
): Promise<MicroReportResponse> {
  const { data } = await apiClient.post<MicroReportResponse>(
    `/diagnostic/result/${token}/micro-diagnostic/${microId}/finish`
  );
  return data;
}

/** Normaliza e-mail para comparação (trim + minúsculas). */
function normalizeEmail(email: string): string {
  return email.trim().toLowerCase();
}

export async function checkExistingDiagnostic(email: string) {
  const { data } = await apiClient.get<{
    exists: boolean;
    diagnostic_id?: string;
    status?: string;
    current_phase?: number;
    total_answers?: number;
    started_at?: string;
  }>(`/diagnostic/check-existing`, { params: { email: normalizeEmail(email) } });
  return data;
}

/**
 * Abandona o diagnóstico em andamento (status → abandoned).
 * Permite iniciar um novo diagnóstico com o mesmo e-mail.
 */
export async function abandonDiagnostic(diagnosticId: string): Promise<void> {
  await apiClient.post(`/diagnostic/${diagnosticId}/abandon`);
}

File: nara-backend/app/api/v1/diagnostic.py
"""Endpoints de diagnóstico."""
import logging
from fastapi import APIRouter, HTTPException, Query, Request, status
from pydantic import EmailStr

from app.models.diagnostic import (
    AnswerSubmitRequest,
    AnswerSubmitResponse,
    DiagnosticStartRequest,
    DiagnosticStartResponse,
    EligibilityResponse,
    DiagnosticResultResponse,
    NextQuestionsResponse,
    QuestionResponse,
)
from app.core.rate_limit import limiter
from app.services.diagnostic_service import DiagnosticService

router = APIRouter()
service = DiagnosticService()
logger = logging.getLogger(__name__)


@router.post("/start", response_model=DiagnosticStartResponse)
@limiter.limit("10/minute")
async def start_diagnostic(request: Request, payload: DiagnosticStartRequest):
    """
    Inicia um novo diagnóstico.
    Requer email e consentimento de privacidade.
    Retorna as 15 perguntas da Fase 1 (Baseline) e token para acesso ao resultado.
    """
    if not payload.consent_privacy:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Consentimento de privacidade é obrigatório",
        )

    result = await service.start_diagnostic(
        email=payload.email,
        full_name=payload.full_name,
        session_id=payload.session_id,
        consent_privacy=payload.consent_privacy,
        consent_marketing=payload.consent_marketing,
        device_info=payload.device_info,
        utm_source=payload.utm_source,
    )

    # Garantir que questions tenham o formato esperado pelo response model
    questions = [
        QuestionResponse(
            id=q.get("id", i),
            area=q.get("area", "Geral"),
            type=q.get("type", "open_long"),
            text=q.get("text", ""),
            scale_labels=q.get("scale_labels"),
            follow_up_hint=q.get("follow_up_hint"),
        )
        for i, q in enumerate(result["questions"], 1)
    ]

    return DiagnosticStartResponse(
        diagnostic_id=result["diagnostic_id"],
        status=result["status"],
        phase=result["phase"],
        questions=questions,
        total_questions=result["total_questions"],
        result_token=result["result_token"],
    )


@router.post("/{diagnostic_id}/answer", response_model=AnswerSubmitResponse)
@limiter.limit("120/minute")
async def submit_answer(request: Request, diagnostic_id: str, payload: AnswerSubmitRequest):
    """
    Submete uma resposta para uma pergunta.
    Atualiza contadores e scores parciais; retorna progresso e elegibilidade.
    """
    try:
        result = await service.submit_answer(
            diagnostic_id=diagnostic_id,
            question_id=payload.question_id,
            question_text=payload.question_text,
            question_area=payload.question_area,
            answer_text=payload.answer_text,
            answer_scale=payload.answer_scale,
            response_time_seconds=payload.response_time_seconds,
        )
        return AnswerSubmitResponse(**result)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        )
    except Exception as e:
        logger.exception("submit_answer failed: %s", e)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Erro interno do servidor. Tente novamente.",
        )


@router.get("/{diagnostic_id}/next-questions", response_model=NextQuestionsResponse)
@limiter.limit("20/minute")
async def get_next_questions(request: Request, diagnostic_id: str):
    """
    Gera perguntas para a próxima fase (RAG + LLM).
    Tempo típico de geração: 3-5 segundos (pode chegar a ~30s em carga).
    """
    try:
        result = await service.get_next_questions(diagnostic_id)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e) if str(e) else "Diagnóstico não encontrado",
        )
    except Exception as e:
        logger.exception("Erro ao gerar próximas perguntas: %s", e)
        msg = str(e) if str(e) else "Falha ao gerar perguntas (RAG/LLM). Tente novamente em instantes."
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=msg,
        )
    questions = [
        QuestionResponse(
            id=q.get("id", i),
            area=q.get("area", "Geral"),
            type=q.get("type", "open_long"),
            text=q.get("text", ""),
            scale_labels=q.get("scale_labels"),
            follow_up_hint=q.get("follow_up_hint"),
        )
        for i, q in enumerate(result["questions"], 1)
    ]
    return NextQuestionsResponse(
        phase=result["phase"],
        questions=questions,
        total_questions=result["total_questions"],
    )


@router.get("/{diagnostic_id}/eligibility", response_model=EligibilityResponse)
@limiter.limit("60/minute")
async def check_eligibility(request: Request, diagnostic_id: str):
    """
    Verifica se o diagnóstico pode ser finalizado.
    Critérios: ≥40 perguntas OU ≥3500 palavras e ≥12 áreas cobertas.
    """
    result = await service.check_eligibility(diagnostic_id)
    return EligibilityResponse(**result)


@router.post("/{diagnostic_id}/finish", response_model=DiagnosticResultResponse)
@limiter.limit("10/minute")
async def finish_diagnostic(request: Request, diagnostic_id: str):
    """
    Finaliza o diagnóstico e gera o relatório.
    Valida elegibilidade antes; geração via RAG + LLM (5-10 s).
    """
    eligibility = await service.check_eligibility(diagnostic_id)
    if not eligibility["can_finish"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Diagnóstico não atende aos critérios. Progresso: {eligibility['overall_progress']:.1f}%",
        )
    result = await service.finish_diagnostic(diagnostic_id)
    return DiagnosticResultResponse(**result)


@router.get("/check-existing")
@limiter.limit("30/minute")
async def check_existing_diagnostic(
    request: Request,
    email: EmailStr = Query(..., description="Email do titular"),
):
    """
    Verifica se existe diagnóstico em andamento para o email.
    Usado na tela de início para oferecer retomada.
    """
    return await service.check_existing_diagnostic(email)


@router.get("/{diagnostic_id}/current-state")
@limiter.limit("60/minute")
async def get_current_state(request: Request, diagnostic_id: str):
    """
    Retorna o estado atual do diagnóstico para retomada (perguntas restantes, progresso).
    """
    try:
        return await service.get_current_state(diagnostic_id)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        )


# Declarar antes de /{diagnostic_id}/result para que /result/{token} não seja capturado como diagnostic_id
@router.get("/result/{token}", response_model=DiagnosticResultResponse)
@limiter.limit("30/minute")
async def get_result_by_token(request: Request, token: str):
    """Obtém resultado pelo token público (sem autenticação)."""
    result = await service.get_result_by_token(token)
    if result is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Resultado não encontrado ou token inválido.",
        )
    return DiagnosticResultResponse(**result)


@router.get("/{diagnostic_id}/result", response_model=DiagnosticResultResponse)
@limiter.limit("30/minute")
async def get_result(request: Request, diagnostic_id: str):
    """Obtém o resultado de um diagnóstico finalizado."""
    result = await service.get_result(diagnostic_id)
    if result is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Resultado não encontrado. Diagnóstico pode não ter sido finalizado.",
        )
    return DiagnosticResultResponse(**result)


@router.post("/{diagnostic_id}/send-resume-link")
@limiter.limit("5/minute")
async def send_resume_link(request: Request, diagnostic_id: str):
    """
    Envia email com link para retomar o diagnóstico.
    Usado quando o usuário clica em "Sair e continuar depois".
    """
    from app.services.email_service import email_service
    from app.database import supabase
    
    # Buscar diagnóstico
    diag_result = supabase.table("diagnostics").select("*").eq(
        "id", diagnostic_id
    ).single().execute()
    
    if not diag_result.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Diagnóstico não encontrado"
        )
    
    diagnostic = diag_result.data
    
    # Calcular progresso
    total_answers = diagnostic.get("total_answers", 0)
    progress = min(100, int((total_answers / 40) * 100))
    
    # Enviar email
    try:
        await email_service.send_resume_link(
            to=diagnostic["email"],
            user_name=diagnostic.get("full_name"),
            diagnostic_id=diagnostic_id,
            progress=progress
        )
        
        return {
            "status": "sent",
            "message": f"Email enviado para {diagnostic['email']}"
        }
    except Exception as e:
        logger.exception("send_resume_link failed for diagnostic_id=%s: %s", diagnostic_id, e)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Erro ao enviar email de retomada. Tente novamente."
        )
